### PR TITLE
Action to create, test and publish wheels

### PR DESCRIPTION
**Pull request by CaselIT**
# Summary of Changes

Action to generate the wheels on release. Fixes https://github.com/falconry/falcon/issues/1281

Track prograss:
- https://github.com/CaselIT/falcon/releases
- https://github.com/CaselIT/falcon/actions/runs/201295934
- https://test.pypi.org/project/falcon-caselit/3.0.0.dev5/#files

TODO:
- [ ] remove various todos
- [x] create upload sdist
- [x] create c files to place in sdist
- [x] crate wheels for linux (manylinux1,2010,2014)
- [x] create wheels for windows & osx
- [x] create wheels for other supported platforms
- [x] secrets must be added by the falcon maintainers to allow pushing to pypi (also on the test pypi I would think?)

### CODE REVIEW

**Comment by CaselIT** [comment]
Regarding the compilation of cython files with cython, are the generated c files always the same regardless of the python 3 version?
I generate them with `cythonize -3 falcon/**/*.pyx`, but I'm not sure if it is ok

**Comment by CaselIT** [comment]
For arm64, there are problems testing for python 3.5, since it tries to install `maturin` from source, and it requires rust to be installed. I think we could not provide a wheel for py 3.5 arm64. Or we skip testing it?

File: setup.cfg [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
+#     ignore:Using or importing the ABCs:DeprecationWarning
+#     ignore:cannot collect test class 'TestClient':pytest.PytestCollectionWarning
+
+env =
+    FALCON_ASGI_WRAP_NON_COROUTINES = Y
+    FALCON_TESTING_SESSION = Y
+    PYTHONASYNCIODEBUG = 1
+
+[flake8]
+max-line-length = 100
```
Comment (CaselIT) [inline setup.cfg:22]: This was a mistake in stashing. it will be removed once I revert the `tag_build`

**Comment by kgriffs** [comment]
This is looking good! I added a token to the project secrets for the real index. Re the test index, I am waiting on an inquiry to see if we can take over https://test.pypi.org/project/falcon/ from the current owner.

**Comment by kgriffs** [comment]
Question: Is there a convenient way to run the action in "test mode" which would flip it to point to the test index and disable the upload of the release assets? I am admittedly not familiar with GH actions to know if it would be easy to toggle this...

**Comment by CaselIT** [comment]
> 
> 
> Question: Is there a convenient way to run the action in "test mode" which would flip it to point to the test index and disable the upload of the release assets? I am admittedly not familiar with GH actions to know if it would be easy to toggle this...

Not currently. I think it could be done by supporting a manual trigger and do the steps that upload the wheel to github and to pypi only if the triggering event is a release. 

Regarding changing the test pypi I don't know, since you need to both change the url and the token, since pypi and the test pipy are completely separate installation and do not share the accounts.
I guess that the `TWINE_REPOSITORY_URL` could be set to a github secret, so that the configuration would be:
- to use the official pypi: official token, remove the secret connected to the url.
- to use the test pypi: test token, create a secret connected to the `TWINE_REPOSITORY_URL` with the test url as value.

Another alternative (but I'm not sure if it's possible) is to add an if to the upload to pypi step to check for a particular message in the release message to disable it on demand?

**Comment by kgriffs** [comment]
[Added TEST_PYPI_TOKEN secret]

**Comment by CaselIT** [comment]
Ok, I'll do the update to the pr

**Comment by CaselIT** [comment]
Updated. IIRC to test this this workflow should be on the default branch, so it's probably best to test on a fork, instead of the actual repo here.
The action trigger when a release (not a tag) is created.

I think we could keep the wheel creation targeted to test pypi for 1 or 2 of the alpha/beta release, check that everything works, then manually upload the wheel to the live pypi for that couple of test run. Then a follow on pr (maybe before the last rc) can change it target the actual pypi.

File: setup.cfg [hunk 2]
```diff
@@ -13,4 +11,4 @@ filterwarnings =
     ignore:Unknown REQUEST_METHOD. '(FOO|BAR|BREW|SETECASTRONOMY)':wsgiref.validate.WSGIWarning
     ignore:"@coroutine" decorator is deprecated:DeprecationWarning
     ignore:Using or importing the ABCs:DeprecationWarning
-    ignore:cannot collect test class 'TestClient':pytest.PytestCollectionWarning
+    ignore:cannot collect test class 'TestClient':pytest.PytestCollectionWarning
```
Comment (vytas7) [inline setup.cfg:14]: No newline at the end of file.

**Review by vytas7 — CHANGES_REQUESTED** [review]
As far as I am qualified (I haven't been involved in the Falcon release process too much), this looks great :+1: Bar one missing newline at the end of file, this is ready to go if @kgriffs is happy.

Btw, do `setup.cfg` additions also address https://github.com/falconry/falcon/issues/1652?

File: setup.cfg (hunk 2 above)
Comment (CaselIT) [reply setup.cfg]: done.

**Comment by CaselIT** [comment]
> Btw, do setup.cfg additions also address #1652? nvm I misread, you only added these to the respective envs.

That was the setup.cfg I use locally, since I like using pytest directly wrt tox. They indeed fix #1652 
They where published because I mistakenly stages all the file instead of only the relevant lines

**Comment by vytas7** [comment]
> They where published because I mistakenly stages all the file instead of only the relevant lines.

That explains. Because I swear I saw them! :slightly_smiling_face:

**Review by vytas7 — APPROVED** [review]
:100:

**Comment by CaselIT** [comment]
It would be great if we could review and merge this, so we can tweak it in the next pre-releases

**Review by vytas7 — APPROVED** [review]

**Review by kgriffs — APPROVED** [review]

**Comment by CaselIT** [comment]
Thanks for merging. I've made a follow up at #1787 that fixes an error with linux build that has come up after I created this

**Pull request merged**